### PR TITLE
oboe: add ErrorOrValue dual return

### PR DESCRIPTION
**Pull request by philburk**
For returning an error code and a frame count, or other value, at the same time.

### CODE REVIEW

File: include/oboe/ErrorOrValue.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
```
Comment (mnaganov) [inline include/oboe/ErrorOrValue.h]: 2018

File: include/oboe/ErrorOrValue.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        return mError;
+    }
+
+    T value() const {
+        return mValue;
+    }
+
+    /**
+     * Quick way to check for an error.
+     * @return true if an error occurred  // TODO does this seem backwards?
```
Comment (mnaganov) [inline include/oboe/ErrorOrValue.h]: To me, "backwards" seems more natural:

`if (!write(...)) { // handle error }`

Thus, operator! should return 'true' if there was an error.

File: include/oboe/ErrorOrValue.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+template <typename T>
+class ErrorOrValue {
+public:
+    explicit ErrorOrValue(oboe::Result error)
+            : mValue{}
+            , mError(error) {}
+
+    explicit ErrorOrValue(T value)
+            : mValue(value < 0 ? 0 : value)
```
Comment (mnaganov) [inline include/oboe/ErrorOrValue.h]: That's not generic. What if 'T' is bool, or a string? In fact, it's the responsibility of the function that returns ErrorOrValue to determine the error condition, and use the appropriate type to pass to the constructor. If the constructor that takes 'T' is used, ErrorOrValue should assume that it was a success.

File: include/oboe/ErrorOrValue.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            , mError(value < 0 ? static_cast<Result>(value) : oboe::Result::OK) {}
+
+    oboe::Result error() const {
+        return mError;
+    }
+
+    T value() const {
+        return mValue;
```
Comment (mnaganov) [inline include/oboe/ErrorOrValue.h]: Should check if there is no error, and emit a log message.

File: src/aaudio/AudioStreamAAudio.cpp [hunk 5]
```diff
@@ -282,25 +282,29 @@ Result AudioStreamAAudio::requestStop() {
...
-int32_t AudioStreamAAudio::write(const void *buffer,
+ErrorOrValue<int32_t>   AudioStreamAAudio::write(const void *buffer,
                                      int32_t numFrames,
                                      int64_t timeoutNanoseconds) {
     AAudioStream *stream = mAAudioStream.load();
     if (stream != nullptr) {
-        return mLibLoader->stream_write(mAAudioStream, buffer, numFrames, timeoutNanoseconds);
+        int32_t result = mLibLoader->stream_write(mAAudioStream, buffer,
+                                                  numFrames, timeoutNanoseconds);
+        return ErrorOrValue<int32_t>(result);
```
Comment (mnaganov) [inline src/aaudio/AudioStreamAAudio.cpp]: As I've noted, it's the responsibility of `AudioStreamAAudio::write` to implement the logic that considers whether the value of `result` indicates an error or not. `ErrorOrValue` must not know these details.

File: include/oboe/ErrorOrValue.h (hunk 1 above)
Comment (philburk) [reply include/oboe/ErrorOrValue.h]: Done

File: include/oboe/ErrorOrValue.h (hunk 2 above)
Comment (philburk) [reply include/oboe/ErrorOrValue.h]: Done

File: include/oboe/ErrorOrValue.h (hunk 3 above)
Comment (philburk) [reply include/oboe/ErrorOrValue.h]: Done

File: src/aaudio/AudioStreamAAudio.cpp (hunk 5 above)
Comment (philburk) [reply src/aaudio/AudioStreamAAudio.cpp]: Done

File: include/oboe/ErrorOrValue.h (hunk 4 above)
Comment (philburk) [reply include/oboe/ErrorOrValue.h]: Why?

Comment (mnaganov) [reply include/oboe/ErrorOrValue.h]: Might be useful if a programmer claims that "oboe doesn't work," and then from the logs it turns out they are not checking the result, and there was an error. Treble for example has even stronger checks where retrieving the value crashes if the error status hasn't been checked before value retrieval. But that feels a bit too harsh, a simple log should be enough.

Comment (philburk) [reply include/oboe/ErrorOrValue.h]: @mnaganov - I do not understand this request. Do you want to log every access to the value?!

Comment (mnaganov) [reply include/oboe/ErrorOrValue.h]: Only if there was an error (mError != OK). Accessing the value if there was an error indicates a problem with client code.

File: src/opensles/AudioStreamBuffered.cpp [hunk 6]
```diff
@@ -98,23 +98,23 @@ int64_t AudioStreamBuffered::predictNextCallbackTime() {
 // TODO: Consider returning an error_or_value struct instead.
```
Comment (dturner) [inline src/opensles/AudioStreamBuffered.cpp]: Nit: Remove this comment now method returns ErrorOrValue

File: include/oboe/ErrorOrValue.h (hunk 4 above)
Comment (philburk) [reply include/oboe/ErrorOrValue.h]: I see. Good idea. https://github.com/google/oboe/issues/76

**Pull request merged**